### PR TITLE
 SUS-3455: Create Special:ListGlobalUsers

### PR DESCRIPTION
**Pull request by mszabo-wikia**
Create a special page that allows to list members of global groups.

https://wikia-inc.atlassian.net/browse/SUS-3455

~**Don't merge** - waiting for translations~

### CODE REVIEW

File: extensions/wikia/ListGlobalUsers/DatabaseGlobalUsersService.php [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+		$dbr = wfGetDB( DB_SLAVE, [], $wgExternalSharedDB );
+
+		$selectQuery = [ 'ug_group' => $groupSet ];
+		$userIds = $dbr->selectFieldValues( 'user_groups', 'ug_user', $selectQuery, __METHOD__, [ 'DISTINCT' ] );
+
+		$userInfo = [];
+
+		foreach ( $userIds as $userId ) {
+			$userInfo[$userId] = User::whoIs( $userId );
+		}
```
Comment (macbre) [inline extensions/wikia/ListGlobalUsers/DatabaseGlobalUsersService.php]: `User::whoAre` should make just a single query for all users.

File: extensions/wikia/ListGlobalUsers/ListGlobalUsers.setup.php [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+$GLOBALS['wgAutoloadClasses']['SpecialListGlobalUsersController'] = __DIR__ . '/SpecialListGlobalUsersController.php';
+
+$GLOBALS['wgSpecialPages']['ListGlobalUsers'] = 'SpecialListGlobalUsersController';
+$GLOBALS['wgSpecialPageGroups']['ListGlobalUsers'] = 'users';
+
+// aliases
+$GLOBALS['wgSpecialPages']['ListStaff'] = 'SpecialListStaff';
+$GLOBALS['wgSpecialPages']['ListVstf'] = 'SpecialListVstf';
+$GLOBALS['wgSpecialPages']['ListHelpers'] = 'SpecialListHelpers';
```
Comment (macbre) [inline extensions/wikia/ListGlobalUsers/ListGlobalUsers.setup.php:19]: :+1:

File: extensions/wikia/ListGlobalUsers/SpecialListGlobalUsersController.php [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+class SpecialListGlobalUsersController extends WikiaSpecialPageController {
+	use PermissionsServiceAccessor;
+
+	public function __construct() {
+		parent::__construct( 'ListGlobalUsers' );
+	}
+
+	public function index() {
+		$this->setHeaders();
+		$this->response->setFormat( WikiaResponse::FORMAT_HTML );
```
Comment (macbre) [inline extensions/wikia/ListGlobalUsers/SpecialListGlobalUsersController.php:22]: Do we need to set a format here?

**Review by macbre — APPROVED** [review]

**Review by macbre — APPROVED** [review]

**Review by macbre — APPROVED** [review]

**Review by macbre — APPROVED** [review]

File: extensions/wikia/ListGlobalUsers/DatabaseGlobalUsersService.php [hunk 4]
```diff
@@ -10,15 +10,28 @@ public function getGroupMembers( array $groupSet ): array {
...
+		$userIdsToGroups = [];
+
+		foreach ( $result as $row ) {
+			$userIdsToGroups[$row->ug_user] = $userIdsToGroups[$row->ug_user] ?? [];
+
+			$userIdsToGroups[] = $row->ug_group;
+			$userIds[] = $row->ug_user;
+		}
+
+		$userNameMap = User::whoAre( $userIds );
```
Comment (macbre) [inline extensions/wikia/ListGlobalUsers/DatabaseGlobalUsersService.php]: :+1:

**Review by macbre — APPROVED** [review]

File: extensions/wikia/ListGlobalUsers/DatabaseGlobalUsersService.php [hunk 5]
```diff
@@ -0,0 +1,41 @@
+<?php
+
+class DatabaseGlobalUsersService implements GlobalUsersService {
+	public function getGroupMembers( array $groupSet ): array {
+		if ( empty( $groupSet ) ) {
+			return [];
+		}
+
+		global $wgExternalSharedDB;
+		$dbr = wfGetDB( DB_SLAVE, [], $wgExternalSharedDB );
```
Comment (pwojcik86) [inline extensions/wikia/ListGlobalUsers/DatabaseGlobalUsersService.php:10]: I'm not aware of MW conventions, but wouldn't it be better to pass the db handle in the constructor, or at least as an argument to this function?

File: extensions/wikia/ListGlobalUsers/modules/ext.wikia.ListGlobalUsers.js [hunk 6]
```diff
@@ -0,0 +1,26 @@
+/* global jQuery, mediaWiki */
+(function ($, mw) {
+	'use strict';
+
+	var areAllSelected = false,
+		selectToggle = document.getElementById('list-global-users-selector-toggle'),
```
Comment (pwojcik86) [inline extensions/wikia/ListGlobalUsers/modules/ext.wikia.ListGlobalUsers.js:6]: vars will look better aligned

File: extensions/wikia/ListGlobalUsers/templates/SpecialListGlobalUsers_index.php [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+				<label for="<?= Sanitizer::encodeAttribute( $checkBox['groupId'] ); ?>"><?= htmlspecialchars( $checkBox['groupLabel'] ); ?></label>
+			</div>
+		<?php endforeach; ?>
+		</div>
+		<input type="submit" class="list-global-users-submit wds-button wds-is-squished" value="<?= wfMessage( 'htmlform-submit' )->escaped(); ?>" />
+	</fieldset>
+</form>
+<ul class="list-global-users-members">
+	<?php foreach ( $userMap as $userId => $userData ): ?>
+		<li><?= Linker::userLink( $userId, $userData['name'] ) . Linker::userToolLinks( $userId, $userData['name'],
```
Comment (pwojcik86) [inline extensions/wikia/ListGlobalUsers/templates/SpecialListGlobalUsers_index.php:22]: This seems like too much code for a view template

File: skins/shared/styles/buttons/buttons-wikia.scss [hunk 8]
```diff
@@ -8,11 +8,11 @@ These buttons are used everywhere on our wikis.
 a.wikia-button,
 .wikia-single-button a,
 .wikia-menu-button,
-input[type="submit"],
-input[type="reset"],
-input[type="button"],
-.button,
-button {
+input[type="submit"]:not(.wds-button),
```
Comment (pwojcik86) [inline skins/shared/styles/buttons/buttons-wikia.scss:11]: Is this change related to the new special page?

**Review by pwojcik86 — APPROVED** [review]
left some comments, but overall lgtm

File: skins/shared/styles/buttons/buttons-wikia.scss (hunk 8 above)
Comment (mszabo-wikia) [reply skins/shared/styles/buttons/buttons-wikia.scss:11]: Yeah, otherwise 2010 era button styles will conflict with Design System buttons

Comment (HM100) [reply skins/shared/styles/buttons/buttons-wikia.scss]: Alright, I'm doing that change to my 2018 A CSS sitewide pack too to avoid conflicts.

Comment (mszabo-wikia) [reply skins/shared/styles/buttons/buttons-wikia.scss:11]: this isn't really relevant to this conversation

Comment (HM100) [reply skins/shared/styles/buttons/buttons-wikia.scss:11]: I know but thanks for fixing them so as not to conflict them, Because if that didn't happened, buttons would have left the gradients.

Comment (HM100) [reply skins/shared/styles/buttons/buttons-wikia.scss:11]: Now after the update is live, the sign out button is now bugged so this needs to be fixed too. Moreover, it also break some button stylings done on some wikis too.
